### PR TITLE
kubeadm: dual-stack validation allow single stack

### PR DESCRIPTION
**Pull request by aojea**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug
/kind cleanup

> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

It turns out that the dual-stack feature enabled doesn't mean that
the cluster MUST be dual-stack, it only indicates that it MAY be
dual-stack but CAN be single-stack.

We should relax the validation to allow single-stack clusters
with dual-stack enabled.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

xref: https://github.com/kubernetes/kubernetes/pull/86895

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm allows to configure single-stack clusters if dual-stack is enabled
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
/assign @neolit123 @liggitt @smarterclayton 
is this the approach we are going to take with dual-stack?
dual-stack with > 2 subnets MUST be dual-stack (at least 1 cidr of each ip Family and only those will be used)
dual-stack may be single stack if only 1 subnet provided

same logic applied to kube-proxy here https://github.com/kubernetes/kubernetes/pull/87353
/hold until this has a wider consensus

**Comment by liggitt** [comment]
I'll defer to sig-network
/assign @thockin

**Comment by neolit123** [comment]
/assign @rosti @Arvinderpal 
for kubeadm

**Comment by Arvinderpal** [comment]
IMO, we should keep the kubeadm logic as it is today.  Having a hard requirement that `features.IPv6DualStack=true` must be accompanied with >1 pod CIDR blocks reduces any confusions that may arise.
The user can always disable feature-gate by setting it to `true/false` based on if they want single or dual stack.

**Comment by liggitt** [comment]
> IMO, we should keep the kubeadm logic as it is today. Having a hard requirement that `features.IPv6DualStack=true` must be accompanied with >1 pod CIDR blocks reduces any confusions that may arise.
> The user can always disable feature-gate by setting it to `true/false` based on if they want single or dual stack.

Feature gates are not long-term configuration options. They are intended to progress to beta, be enabled by default, progress to GA, be locked on, and then be removed. Feature-gate enablement means the code related to that feature is active. In this case, additional configuration (a second service cidr in a second family) is required to have a dual-stack cluster.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go
Comment (vllry) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go]: Should also account for >1 CIDR in non-dualstack mode.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (aojea) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: we should decide if we allow more than 2 and we use only the first from each family, `len(subnets) > 1`
or we just only allow 2, 1 from each family ' len(subnets) ==2`

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (aojea) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: that means that it will validate subnets despite we already know that the config is invalid 
`!isDualStack && len(subnets) > 1:`
is that ok?

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (liggitt) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: allowing more than two and arbitrarily skipping/ignoring some is really confusing… that doesn't seem like a good idea

**Comment by Arvinderpal** [comment]
> > IMO, we should keep the kubeadm logic as it is today. Having a hard requirement that `features.IPv6DualStack=true` must be accompanied with >1 pod CIDR blocks reduces any confusions that may arise.
> > The user can always disable feature-gate by setting it to `true/false` based on if they want single or dual stack.
> 
> Feature gates are not long-term configuration options. They are intended to progress to beta, be enabled by default, progress to GA, be locked on, and then be removed. Feature-gate enablement means the code related to that feature is active. In this case, additional configuration (a second service cidr in a second family) is required to have a dual-stack cluster.

I'm fine with this change. However,I was not implying that feature gate here should be a long-term configuration. When I wrote the initial code, I wanted to make sure there was a hard separation between single and dual stack at least for the early stages, so as to not screw up the primary use of a singel stack cluster.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (rosti) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Yes, that's fine. The idea here is to give as much possible errors to the user as possible. That way they can fix most (if not all) at once. This UX is better than having users fix one error, then re-run kubeadm to get another and so on.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (rosti) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Let's allow for two (but no more).

**Review by rosti — COMMENTED** [review]
Thanks @aojea !

**Comment by aojea** [comment]
@rosti @vllry PTAL

**Comment by aojea** [comment]
/retest 
known flakes

**Comment by aojea** [comment]
/retest
1 storage test flake

**Comment by aojea** [comment]
/retest
```
[sig-apps] CronJob should delete successful finished jobs with limit of one successful job expand_more	3m36s
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: inline ephemeral CSI volume] ephemeral should create read-only inline ephemeral volume expand_more
```

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (rosti) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Slight wording here:
```"expected one (IPv4 or IPv6) CIDR or two CIDRs from each family for dual-stack networking"```
[1]

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (rosti) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Probably it makes more sense to have here the same message as in [1]

**Review by rosti — COMMENTED** [review]
Thanks @aojea !
We are on the final stretch here!

**Review by rosti — APPROVED** [review]
Superb! Thanks for the speedy reaction @aojea !
/approve
/lgtm
/hold cancel

**Comment by aojea** [comment]
/retest
```
[sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: tmpfs] [Testpattern: Pre-provisioned PV (default fs)] subPath should support existing directories when readOnly specified in the volumeSource expand_more	2m1s
[sig-api-machinery] Garbage collector should delete RS created by deployment when not orphaning [Conformance] expand_more	1m31s
[sig-apps] DisruptionController evictions: maxUnavailable deny evictions, integer => should not allow an eviction expand_more
```

**Pull request merged**
